### PR TITLE
when cap is enough, reuse slice data for compatibility with encoding/json

### PR DESCRIPTION
**Pull request by IncSW**
```
import "encoding/json "

func TestJSONSlice(t *testing.T) {
	slice := []int{}
	err := json.Unmarshal([]byte(`[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]`), &slice)
	if err != nil {
		t.Fatal(err)
	}
	t.Log(
		len(slice), // 10
		cap(slice), // 13,
	)

	err = json.Unmarshal([]byte(`[11]`), &slice)
	if err != nil {
		t.Fatal(err)
	}
	t.Log(
		len(slice), // 1
		cap(slice), // 13
		slice[0],   // 11
		*(*int)(unsafe.Pointer((*reflect.SliceHeader)(unsafe.Pointer(&slice)).Data + 8*9)), // 10
	)
}
```

### CODE REVIEW

File: decode_slice.go [hunk 1]
```diff
@@ -96,10 +96,11 @@ func (d *sliceDecoder) decodeStream(s *stream, depth int64, p unsafe.Pointer) er
...
 			if s.char() == ']' {
-				*(*sliceHeader)(p) = sliceHeader{
-					data: newArray(d.elemType, 0),
-					len:  0,
-					cap:  0,
+				dst := (*sliceHeader)(p)
+				if dst.data == nil {
+					dst.data = newArray(d.elemType, 0)
+				} else {
+					dst.len = 0
```
Comment (goccy) [inline decode_slice.go:103]: 👍

File: decode_slice.go [hunk 2]
```diff
@@ -130,11 +131,11 @@ func (d *sliceDecoder) decodeStream(s *stream, depth int64, p unsafe.Pointer) er
...
-					dstCap := idx + 1
-					dst := sliceHeader{
-						data: newArray(d.elemType, dstCap),
-						len:  idx + 1,
-						cap:  dstCap,
+					dst := *(*sliceHeader)(p)
+					dst.len = idx + 1
+					if dst.len > dst.cap {
+						dst.data = newArray(d.elemType, dst.len)
+						dst.cap = dst.len
```
Comment (goccy) [inline decode_slice.go:138]: LGTM 👍 
But  It seems that `L145:*(*sliceHeader)(p) = dst` is not needed .

**Review by goccy — COMMENTED** [review]
Thank you for your great PR. I commented a little.

**Comment by goccy** [comment]
LGTM !!! Thank you for the contribution !

**Comment by IncSW** [comment]
LGTM. Thank you for the contribution.

**Pull request merged**